drivers/led/lp50xx.c

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ * Copyright (c) 2022 Grinn
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief LP50xx LED controller
+ */
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/drivers/led/lp50xx.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(lp50xx, CONFIG_LED_LOG_LEVEL);
+
+#define LP50XX_MIN_BRIGHTNESS		0U
+#define LP50XX_MAX_BRIGHTNESS		100U
+
+#define LP5012_MAX_MODULES	4
+#define LP5024_MAX_MODULES	8
+#define LP5036_MAX_MODULES	12
+
+
+/* Base registers */
+#define LP50XX_DEVICE_CONFIG0		0x00
+#define LP50XX_DEVICE_CONFIG1		0x01
+#define LP50XX_LED_CONFIG0		0x02
+
+#define LP50XX_BASE_REG(chip)		((chip) < LP5030 ? 0x03 : 0x04)
+
+/* Register values */
+#define CONFIG0_CHIP_EN			BIT(6)
+
+#define CONFIG1_LED_GLOBAL_OFF		BIT(0)
+#define CONFIG1_MAX_CURRENT_OPT		BIT(1)
+#define CONFIG1_PWM_DITHERING_EN	BIT(2)
+#define CONFIG1_AUTO_INCR_EN		BIT(3)
+#define CONFIG1_POWER_SAVE_EN		BIT(4)
+#define CONFIG1_LOG_SCALE_EN		BIT(5)
+
+/* Chipset models supported by this driver. */
+enum lp50xx_chips {
+	LP5009,
+	LP5012,
+	LP5018,
+	LP5024,
+	LP5030,
+	LP5036,
+};
+
+struct lp50xx_config {
+	struct i2c_dt_spec bus;
+	enum lp50xx_chips chip;
+	uint8_t max_modules;
+	uint8_t max_leds;
+	uint8_t num_leds;
+	bool log_scale_en;
+	bool max_curr_opt;
+	const struct led_info *leds_info;
+};
+
+struct lp50xx_data {
+	uint8_t *chan_buf;
+};
+
+static uint8_t lp50xx_get_brightness_base_reg(const struct lp50xx_config *config)
+{
+	return LP50XX_BASE_REG(config->chip) + 4;
+}
```

Comment:
Please come back at your first version. See my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/59852#discussion_r1270848979.

Adding register fields (`brightness_base_reg`, `reset_reg`, etc) to `struct lp50xx_config` was the right approach. It was much better than using runtime switches (as in `lp50xx_get_brightness_base_reg()` above).

Reply:
I saw your previous comment but checking if all registers were necessary made me realize I could calculate offsets based on the number of modules and that was the only information required. So I took this middle ground approach.
I now am able to transform these functions to macros which can be resolved at compile time to populate the fields of the `lp50xx_config` structure. Will that be a better approach ?

Reply:
Yes it will. I think that using the number of "modules" is a great idea. And it is OK to compute offsets at runtime using the register address and number of modules information stored in `struct lp50xx_config`. Let's just avoid the runtime switches on the chip type.